cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/FastMap.scala

```diff
@@ -0,0 +1,428 @@
+}
+
+class FastMapAnyAny[@sp(Int, Long, AnyRef) A, @sp(Int, Long, AnyRef) B]
+  (protected val underlying: Object2ObjectMap[A, B] = new Object2ObjectOpenHashMap[A, B]())
+  extends FastMap[A, B] with JavaUtilMapWrapper[A, B]{
+
+  def +=(key: A, value: B): this.type = {
+    underlying.put(key, value)
+    this
+  }
+
+  def -=(key: A): this.type = {
+    underlying.remove(key)
+    this
+  }
+
+  def get(key: A): B = {
+    underlying.get(key)
+  }
+
+  def contains(key: A) = underlying.containsKey(key)
+
+  private[collections] def iterator: Iterator[(A, B)] = new Iterator[(A, B)] {
+    val it = underlying.object2ObjectEntrySet().iterator()
+
+    override def hasNext: Boolean = it.hasNext
+
+    override def next(): (A, B) = {
+      val next = it.next()
+      (next.getKey, next.getValue)
+    }
+  }
+
+  override protected def utilMap: util.Map[A, B] = underlying
+}
+
+class FastMapAnyInt[@sp(Int, Long, AnyRef) A, MapType <: Object2IntMap[A]](
+    protected val underlying: MapType
+  ) extends FastMap[A, Int] with JavaUtilMapWrapper[A, Int] {
+
+  def +=(key: A, value: Int): this.type = {
+    underlying.put(key, value)
+    this
+  }
+
+  def -=(key: A): this.type = {
+    underlying.remove(key)
+    this
+  }
+
+  def get(key: A): Int = {
+    underlying.getInt(key)
+  }
+
+  override def contains(key: A) = underlying.containsKey(key)
+
+  override private[collections] def iterator: Iterator[(A, Int)] = new Iterator[(A, Int)] {
+    val it = underlying.object2IntEntrySet().iterator()
+
+    override def hasNext: Boolean = it.hasNext
+
+    override def next(): (A, Int) = {
+      val next = it.next()
+      (next.getKey, next.getIntValue)
+    }
+  }
+
+  override protected def utilMap: util.Map[A, Int] = underlying.asInstanceOf[util.Map[A, Int]]
+}
+
+/**
+ * Specialization for `Int -> B` map.
+ */
+class FastMapInt[@sp(Int, Long, AnyRef) B]
+  (private[collections] val underlying: Int2ObjectMap[B] = new Int2ObjectOpenHashMap[B]())
+  extends FastMap[Int, B] {
+
+  override def +=(key: Int, value: B): this.type = {
+    underlying.put(key, value)
+    this
+  }
+
+  override def -=(key: Int): this.type = {
+    underlying.remove(key)
+    this
+  }
+
+  override def get(key: Int): B = {
+    underlying.get(key)
+  }
+
+  override def clear(): Unit = {
+    underlying.clear()
+  }
+
+  override def contains(key: Int) = underlying.containsKey(key)
+
+  override private[collections] def iterator: Iterator[(Int, B)] = new Iterator[(Int, B)] {
+    val it = underlying.int2ObjectEntrySet().iterator()
+
+    override def hasNext: Boolean = it.hasNext
+
+    override def next(): (Int, B) = {
+      val next = it.next()
+      (next.getIntKey, next.getValue)
+    }
+  }
+
+  override def size(): Int = underlying.size()
+}
+
+/**
+ * Specialization for `Int -> Int` map.
+ */
+class FastMapIntInt[MapType <: Int2IntMap]
+  (protected val underlying: MapType)
+  extends FastMap[Int, Int] {
+
+  override def +=(key: Int, value: Int): this.type = {
```

Comment:
It's not possible, because there is no common non-boxed superclass of Int2IntMap and Long2LongMap in fastutils. Thus, we cannot specify the type of underlying. Maybe we could do this with some weird casting, but I would rather avoid that --- I'm afraid that we will have less control over boxing then.